onnx_models/conftest.py

```diff
@@ -20,6 +23,93 @@
 ARTIFACTS_ROOT = THIS_DIR / "artifacts"
 
 
+###############################################################################
+# Configuration
+###############################################################################
+
+
+def pytest_addoption(parser):
+    # List of configuration files following this schema:
+    #   {
+    #     "config_name": str,
+    #     "iree_compile_flags": list of str,
+    #     "iree_run_module_flags": list of str,
+    #     "skip_compile_tests": list of str,
+    #     "skip_run_tests": list of str,
+    #     "tests_and_expected_outcomes": dict
+    #   }
+    #
+    # For example, to run some tests on CPU with the `llvm-cpu` backend and
+    # `local-task` device:
+    #   {
+    #     "config_name": "cpu_llvm_task",
+    #     "iree_compile_flags": ["--iree-hal-target-backends=llvm-cpu"],
+    #     "iree_run_module_flags": ["--device=local-task"],
+    #     "tests_and_expected_outcomes": {
+    #       "default": "skip",
+    #       "tests/foo/bar/baz.py::test_a": "pass",
+    #       "tests/foo/bar/baz.py::test_b": "fail-import",
+    #       "tests/foo/bar/baz.py::test_c": "fail-compile",
+    #       "tests/foo/bar/baz.py::test_d": "fail-run"
+    #     }
+    #   }
+    #
+    # The file can be specified in (by order of preference):
+    #   1. The `--config-file` argument
+    #       e.g. `pytest ... --config-file foo.json`
+    #   2. The `IREE_TEST_CONFIG_FILE` environment variable
+    #       e.g. `export IREE_TEST_CONFIG_FILE=foo.json`
+    #   3. A default config file used for testing the test suite itself
+    default_config_file = os.getenv(
+        "IREE_TEST_CONFIG_FILE", THIS_DIR / "configs" / "onnx_models_cpu_llvm_task.json"
+    )
+    parser.addoption(
+        "--test-config-file",
+        type=Path,
+        default=default_config_file,
+        help="Config JSON file used to parameterize test cases",
+    )
```

Comment:
> * https://github.com/nod-ai/SHARK-TestSuite/blob/main/alt_e2eshark/run.py uses `--device=`, `--backend=`, `--target-chip=`, and `--test-filter=` arguments. Arbitrary flags are not supported, and test expectations are also not supported, so there is no way to directly signal if tests are unexpectedly passing or failing.

Here are some ideas, adding complexity in the conftest file but making it more flexible:

* Add more options here to use instead of `--test-config-file`, like `--iree-compile-flags`
* Make every option a flag then use a flagfile pattern like https://stackoverflow.com/a/27434050 so "config.json" is just a collection of regular flags
  * Not sure how that would work with the lists of tests with expectations... I like having the full list of tests that will run be sorted and not split between groups like in the onnx op tests
* Could load a .py file that has the list of tests and statuses, or even just implements the `pytest_collection_modifyitems` hook